src/libs/ImplicitAuthManager.js

```diff
@@ -0,0 +1,537 @@
+        !TypeCheck.isString(
+          config.redirectURI(ImplicitAuthManager.validAPIIntentions.LOGIN)
+        )
+      ) {
+        throw new Error(
+          'If passing in a custom redirectURI as a function it must return a string'
+        );
+        //otherwise make sure redirectURI is a string
+      } else if (
+        !TypeCheck.isFunction(config.redirectURI) &&
+        !TypeCheck.isString(config.redirectURI)
+      ) {
+        throw new Error(
+          'If passing in a custom redirectURI it must either be a function or a string'
+        );
+      }
+    }
+    //if login URI responseType was passed in
+    if (
+      config.loginURIResponseType &&
+      !TypeCheck.isString(config.loginURIResponseType)
+    ) {
+      throw new Error('loginURIResponseType in config must be typeof [string]');
+    }
+
+    if (
+      config.loginURIResponseType &&
+      !ImplicitAuthManager.validResponseTypes().includes(
+        config.loginURIResponseType
+      )
+    ) {
+      throw new Error(
+        "loginURIResponseType isn't valid, please view https://openid.net/specs/openid-connect-core-1_0.html#ImplicitAuthRequest for details"
+      );
+    }
+    //validate hooks
+    if (config.hooks) {
+      if (!TypeCheck.isObject(config.hooks)) {
+        throw new Error('hooks in config must be typeof [object]');
+      }
+      const validHooks = ImplicitAuthManager.validHooks();
+      //loop over object and validate keys and their type
+      for (let key in config.hooks) {
+        if (!validHooks.includes(key)) {
+          throw new Error(
+            `${key} in config.hooks is not a valid hook, please see API Docs for information on valid hooks`
+          );
+        }
+        if (!TypeCheck.isFunction(config.hooks[key])) {
+          throw new Error(`config.hooks.${key} must be typeof [function]`);
+        }
+      }
+    }
+  }
+
+  createRequestKey() {
+    return CryptoUtils.genCryptographicRandomValue();
+  }
+
+  createNonce() {
+    const requestKey = this.createRequestKey();
+    //save request key in local storage for reference on auth redirect
+    saveDataInLocalStorage('sso', { requestKey });
+    //hash requestKey and return as nonce
+    return CryptoUtils.hashSHA256(requestKey);
+  }
+
+  clearNonce() {
+    deleteDataFromLocalStorage('sso');
+  }
+
+  isAReplayAttack(nonce) {
+    //this could be a replay attack if the nonce contained with the jwt doesn't match
+    //the hashed request key that SHOULD be in local storage
+    const sso = getDataFromLocalStorage('sso');
+    if (TypeCheck.isObject(sso) && sso.requestKey) {
+      return !CryptoUtils.checkAgainstHash(sso.requestKey, nonce);
+    } else {
+      return true;
+    }
+  }
+
+  isTokenExpired(token) {
+    return new Date() / 1000 > token.exp;
+  }
+
+  areTokensExpired() {
+    let expired = 0;
+    //get tokens
+    const tokens = this.getAuthDataFromLocal();
+
+    if (!tokens) {
+      return true;
+    }
+
+    for (let key in tokens) {
```

Comment:
I don't think you need to change this but I'd favour a declerative style rather than imperative eg: `tokens.filter(item => this.isTokenExpired(item)).lenght > 0`.

Reply:
something like `Object.keys(tokens).filter(item => this.isTokenExpired(tokens[item]).length > 0`